### DISCUSSION TITLE
生成自己的token问题

### DISCUSSION THREAD

**Discussion by 786733681** [wechaty/wechaty#2096] [General]
根据 https://github.com/wechaty/wechaty/issues/1986 中的说法
1.拉取镜像，然后设置镜像参数，最终得到的效果是如下图：
![image](https://user-images.githubusercontent.com/39853265/100406723-94d83f80-30a1-11eb-8a88-80ced2459ac6.png)
2.而且我查看了容器wechaty的运行状态，如下图：
![image](https://user-images.githubusercontent.com/39853265/100406793-c81ace80-30a1-11eb-891a-ce1132097130.png)
3.然后验证了一下token,如下图：
![image](https://user-images.githubusercontent.com/39853265/100406943-1203b480-30a2-11eb-97d5-31b71aef9e91.png)
4.最后在项目中set token值，start运行，如下图：
![image](https://user-images.githubusercontent.com/39853265/100407060-5db65e00-30a2-11eb-9161-83f9e10859e6.png)
====================================
结果，项目没有运行起来，按照我的理解，拉取镜像运行后，容器应该是挂在后台的了，但是现在没有，然后token的验证链接那里我随便写了他都是返回{"ip":"0.0.0.0","port":0}，运行的项目的时候连接不上，一直在重连。请问是哪个环节出错，具体是什么错误？
麻烦作者详细说明一下，谢谢！

**Comment by ZakaryCode** [comment]
同样在 token 中遇到过一些问题，不过我可以成功设置自定义的 token，在验证接口中也可以得到成功的验证。

但是项目使用自定义 token 依旧无法成功跑起项目，看了下日志发现自定义 token 的服务会跑一个机器人服务去向官方的服务器请求数据:

- 如果没有设置 token 就会提示 `wechaty-puppet-hostie: token not found (save token to WECHATY_PUPPET_HOSTIE_TOKEN env var or pass it to puppet options is required.). See: <https://github.com/wechaty/wechaty-puppet-hostie#1-wechaty_puppet_hostie_token>`
- 如果设置 token 则会启动机器人服务提供给自定义 token 服务，如果设置错误则和机器人错误 token 得到的反馈一致 `failed to connect to all addresses`

> 目前根据日志理解，所谓自定义 token 应该是一个中转服务，并不能绕过付费 token 服务独立使用（如果是我理解错误希望可以官方可以直接反驳，并给出详细使用方法）
>
> 顺便吐槽下 wechaty 打印的 IO 日志，一直是 `INFO Io on(jsonrpc): [object Object]` 对非官方的开发者十分不友好

**Comment by huan** [comment]
@786733681 It seems that you have not followed the right steps from the #1986 because the logs show that your docker container has exited.

Please check your command of `docker run` and make sure it started successfully.

**Comment by huan** [comment]
@ZakaryCode Glad to know that you can setup your Wechaty Puppet Service successfully and can validate it via our API.

> 目前根据日志理解，所谓自定义 token 应该是一个中转服务，并不能绕过付费 token 服务独立使用（如果是我理解错误希望可以官方可以直接反驳，并给出详细使用方法）

Yes, you are right. The Wechaty Puppet Service can help you publish a token, and you must provide the real service for your token, for example, use an existing token. ;-)

> 顺便吐槽下 wechaty 打印的 IO 日志，一直是 INFO Io on(jsonrpc): [object Object] 对非官方的开发者十分不友好

Pull Request is welcome!

**Comment by 786733681** [comment]
大概明白了，谢谢您的讲解。

**Comment by huan** [comment]
@786733681 you are welcome, and please feel free to let me know if you still have any problems.
